Fixed #26056 -- Added QuerySet.values()/values_list() support for ArrayField's __overlap lookup.

bcail opened the pull request:
This is based on work by Mads Jensen and kosz85 in PR #7838.

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -414,6 +414,24 @@ def test_overlap_charfield_including_expression(self):
         )
 
+    def test_overlap_values_array_field(self):
+        # issue 26056
```
This comment is unnecessary
```suggestion
```

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -414,6 +414,24 @@ def test_overlap_charfield_including_expression(self):
             [obj_1, obj_2],
         )
 
+    def test_overlap_values_array_field(self):
```
```suggestion
    def test_overlap_values_list(self):
```

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -414,6 +414,24 @@ def test_overlap_charfield_including_expression(self):
+        qs = CharArrayModel.objects.filter(
+            field__overlap=CharArrayModel.objects.values_list("field", flat=True)
+        )
+        self.assertSequenceEqual(
```
`qs` is not ordered so we should use `assertCountEqual()`:
```suggestion
        self.assertCountEqual(
```

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -414,6 +414,24 @@ def test_overlap_charfield_including_expression(self):
+            field__overlap=CharArrayModel.objects.values_list("field", flat=True)
+        )
+        self.assertSequenceEqual(
+            qs.values_list("field", flat=True),
```
`flat=True` is unnecessary

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -414,6 +414,24 @@ def test_overlap_charfield_including_expression(self):
 
+    def test_overlap_values_array_field(self):
+        # issue 26056
+        post1 = CharArrayModel(field=["django"])
+        post2 = CharArrayModel(field=["thoughts"])
+        post3 = CharArrayModel(field=["tutorial"])
+        CharArrayModel.objects.bulk_create([post1, post2, post3])
+        qs = CharArrayModel.objects.filter(
+            field__overlap=CharArrayModel.objects.values_list("field", flat=True)
+        )
+        self.assertSequenceEqual(
+            qs.values_list("field", flat=True),
+            [
+                (["django"]),
+                (["thoughts"]),
+                (["tutorial"]),
+            ],
+        )
```
I'd reuse existing objects and add `filter()` to ensure that it actually works, e.g.:
```suggestion
        values = NullableIntegerArrayModel.objects.filter(order__lt=3).values_list(
            "field"
        )
        self.assertCountEqual(
            NullableIntegerArrayModel.objects.filter(
                field__overlap=values,
            ),
            self.objs[:3],
        )
```

felixxm commented on django/contrib/postgres/lookups.py:
```diff
@@ -1,5 +1,6 @@
 from django.db.models import Transform
 from django.db.models.lookups import PostgresOperatorLookup
+from django.db.models.sql.compiler import Query
```
```suggestion
from django.db.models.sql.query import Query
```

felixxm commented on django/contrib/postgres/lookups.py:
```diff
@@ -18,6 +19,14 @@ class Overlap(PostgresOperatorLookup):
     lookup_name = "overlap"
     postgres_operator = "&&"
 
+    def process_rhs(self, qn, connection):
+        if isinstance(self.rhs, Query):
+            from .expressions import ArraySubquery
+
+            self.rhs = ArraySubquery(self.rhs)
+        rhs, params = super().process_rhs(qn, connection)
+        return rhs, params
```
As far as I'm aware, this should be moved to the `get_prep_lookup()`:
```suggestion
    def get_prep_lookup(self):
        from .expressions import ArraySubquery

        if isinstance(self.rhs, Query):
            self.rhs = ArraySubquery(self.rhs)
        return super().get_prep_lookup()
```

felixxm commented on docs/releases/4.2.txt:
```diff
@@ -102,6 +102,9 @@ Minor features
   <django.contrib.postgres.search.TrigramStrictWordDistance>` expressions allow
   using trigram strict word similarity.
 
+* Added support for querying against querysets in
+  :lookup:`arrayfield.overlap`.
```
```suggestion
* :lookup:`arrayfield.overlap` now supports ``QuerySet.values()`` and
  ``values_list()` as right-hand sides.
```

felixxm reviewed (commented):
@bcail Thanks :+1: I left comments. Please also add an example with `QuerySet.values()` on the right-hand side to the `__overlap` [docs](https://docs.djangoproject.com/en/dev/ref/contrib/postgres/fields/#overlap).

bcail commented:
@felixxm Thanks for the review! I've updated the PR.

Note: I updated the tags field for the 'Second Post' in the docs section, to avoid this error: `psycopg2.errors.ArraySubscriptError: cannot accumulate arrays of different dimensionality`.

felixxm commented:
@bcail Thanks for updates :+1: Welcome aboard :boat: 

I wasn't able to push small edits to your branch so I opened #16307.
